helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -93,42 +184,126 @@ public static ClusterModel generateClusterModel(ResourceControllerDataProvider d
+      Map<String, ResourceAssignment> bestPossibleAssignment,
+      Map<String, Set<AssignableReplica>> allocatedReplicas) {
+    Set<AssignableReplica> toBeAssignedReplicas = new HashSet<>();
+    // check each resource to identify the allocated replicas and to-be-assigned replicas.
+    for (String resourceName : replicaMap.keySet()) {
+      // <partition, <state, instances set>>
+      Map<String, Map<String, Set<String>>> baselinePartitionStateMap =
+          getValidStateInstanceMap(baselineAssignment.get(resourceName), activeInstances);
+      Map<String, Map<String, Set<String>>> bestPossiblePartitionStateMap =
+          getValidStateInstanceMap(bestPossibleAssignment.get(resourceName), activeInstances);
+      // Iterate the replicas of the resource to find the ones that require reallocating.
+      for (AssignableReplica replica : replicaMap.get(resourceName)) {
+        String partitionName = replica.getPartitionName();
+        String replicaState = replica.getReplicaState();
+        Set<String> baselineAllocations =
+            baselinePartitionStateMap.getOrDefault(partitionName, Collections.emptyMap())
+                .getOrDefault(replicaState, Collections.emptySet());
+        Set<String> bestPossibleAllocations =
+            bestPossiblePartitionStateMap.getOrDefault(partitionName, Collections.emptyMap())
+                .getOrDefault(replicaState, Collections.emptySet());
+
+        // Compare the best possible assignments with the baseline assignment for the common part.
+        List<String> commonAllocations = new ArrayList<>(bestPossibleAllocations);
+        commonAllocations.retainAll(baselineAllocations);
+        if (!commonAllocations.isEmpty()) {
+          // 1. If the partition is allocated at the same location in both baseline and best possible
+          // assignment, there is no need to reassign it.
+          String allocatedInstance = commonAllocations.get(0);
+          allocatedReplicas.computeIfAbsent(allocatedInstance, key -> new HashSet<>()).add(replica);
+          // Remove the instance from the record to prevent the same location being processed again.
+          baselineAllocations.remove(allocatedInstance);
+          bestPossibleAllocations.remove(allocatedInstance);
+        } else if (!baselineAllocations.isEmpty()) {
+          // 2. If the partition is allocated at an active instance in the Baseline but the
+          // allocation does not exist in the best possible assignment, try to rebalance it.
+          toBeAssignedReplicas.add(replica);
+          // Remove the instance from the baseline record to prevent the same location being picked
+          // up again.
+          baselineAllocations.remove(baselineAllocations.iterator().next());
+        } else if (!bestPossibleAllocations.isEmpty()) {
+          // 3. If the partition is allocated at an active instance in the best possible assignment
+          // only, there is no need to rebalance it.
```

Comment:
Let's also add what's going on conceptually here:
---------------------
This replica exists in the bestPossibleAssignment but does not appear in the baseline. This means that it exists in reality but does not and should not exist because it is not found in the baseline ("ideal/goal state"), implying that the parent resource might have been removed. But this is impossible because if a resource was removed, its IdealState would no longer exist, and this replica shouldn't have been created in this ClusterModel.
So this condition exists as a sanity check and is more about hitting all the possible cases of the if statement than processing the replica, hence the NOP (we just consider it allocated/assigned by passing it through allocatedReplicas).

Reply:
I checked the code again, I was wrong when we talked offline. This condition has been changed multiple times, so I didn't talk with the latest version. So it is really a good point that you mentioned we should note more about the background here.

This code will be triggered if the baseline allocation contains an invalid allocation. For instance, the node is disabled or offline. In this case, we should keep the replica at the same place.
The second point is still valid that we make the conditions complete. But I don't want to put too many comments regarding sanity check since it might distract the reader from understanding the real requirement here. Please double-check my change after I update the PR.